js/ui/payg.js

```diff
@@ -0,0 +1,195 @@
+                                      .format(minutesLeft));
+            } else {
+                this.setErrorMessage(_("Too many attempts. Try again in a few seconds."));
+            }
+
+            // Make sure to clean the status once the time is up (if this dialog is still alive)
+            // and make sure that we install this callback at some point in the future (+1 sec).
+            this._clearTooManyAttemptsId = Mainloop.timeout_add_seconds(Math.max(1, secondsLeft), () => {
+                this.verificationStatus = UnlockStatus.NOT_VERIFYING;
+                this.clearError();
+                this.updateSensitivity();
+                return GLib.SOURCE_REMOVE;
+            });
+
+            this.verificationStatus = UnlockStatus.TOO_MANY_ATTEMPTS;
+            return;
+        }
+
+        // Common errors after this point.
+        if (error.matches(PaygManager.PaygErrorDomain, PaygManager.PaygError.INVALID_CODE)) {
+            this.setErrorMessage(_("Invalid code. Please try again."));
+        } else if (error.matches(PaygManager.PaygErrorDomain, PaygManager.PaygError.CODE_ALREADY_USED)) {
+            this.setErrorMessage(_("Code already used. Please enter a new code."));
+        } else if (error.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.TIMED_OUT)) {
+            this.setErrorMessage(_("Time exceeded while verifying the code"));
+        } else {
+            // We don't consider any other error here (and we don't consider DISABLED explicitly,
+            // since that should not happen), but still we need to show something to the user.
+            this.setErrorMessage(_("Unknown error"));
+        }
+
+        this.verificationStatus = UnlockStatus.FAILED;
+    },
+
+    _onDestroy: function() {
+        if (this._clearTooManyAttemptsId > 0) {
+            Mainloop.source_remove(this._clearTooManyAttemptsId);
+            this._clearTooManyAttemptsId = 0;
+        }
+    },
+
+    setErrorMessage: function(message) {
+        if (message) {
+            this.errorLabel.text = message;
+            this.errorLabel.opacity = 255;
+        } else {
+            this.errorLabel.text = '';
+            this.errorLabel.opacity = 0;
+        }
+    },
+
+    clearError: function() {
+        this.setErrorMessage(null);
+    },
+
+    startSpinning: function() {
+        this.spinner.play();
+        this.spinner.actor.show();
+        Tweener.addTween(this.spinner.actor,
+                         { opacity: 255,
+                           time: SPINNER_ANIMATION_TIME_SECS,
+                           delay: SPINNER_ANIMATION_DELAY_SECS,
+                           transition: 'linear' });
+    },
+
+    stopSpinning: function() {
+        this.spinner.actor.hide();
+        this.spinner.actor.opacity = 0;
+        this.spinner.stop();
+    },
+
+    reset: function() {
+        this.stopSpinning();
+        this.entryReset();
+        this.updateSensitivity();
+    },
+
+    addCodeCallback: function(error) {
```

Comment:
Seems reasonable. I've never gotten into lambdas much but I think this, specifically, was a named function because it was being called in two places at once.

I think my natural inclination is more toward structures that minimize noise in case you make changes later (eg, calling from multiple places) and not doing multiple things in a single line.

But I relocated this as a lambda and it seems fine (and is obviously conventional in this project).